### ISSUE TITLE
opening large files crashes zerobrane

### ISSUE THREAD

**Issue by CapsAdmin** [pkulchenko/ZeroBraneStudio#491]
This is kind of expected and I understand not being able to deal with huge files and all that, but maybe a warning of some sort should be in place before you attempt?

I have some log files I need to view. I think I should probably get log viewer for that specific purpose. Before I used notepad++ but it had limits too if the log was larger than 500 mb or something. The log I tried to open is 311 mb.

**Comment by pkulchenko** [comment]
@CapsAdmin, I think the process simply ran out of memory; at least this is what I see on Windows with a log file of similar size. I agree, the warning would be nice, but it's difficult to estimate how much memory the operation may consume (especially given that some memory is taken by wxwidgets as well). The IDE runs on LuaJIT, so on a 64bit system you may be limited to 1G of memory.

I'll check if the memory consumption may be reduced by not loading the entire file in memory first, but I wouldn't count on this much.

**Comment by madmaxoft** [comment]
I guess checking the file size and comparing it to a fixed threshold would be sufficient in this case? 100 MiB should be a good enough threshold.

**Comment by CapsAdmin** [comment]
I would love to make a package that would let me view log files however. I would need to be able to tell zerobrane what to do before opening a file and then seek and read the file as you scroll or something.

**Comment by pkulchenko** [comment]
@CapsAdmin, I realize the need, but unfortunately this doesn't quite work with Scintilla as it needs to load the entire file to operate on it. It does able to handle large files though. I've made couple of changes to how the IDE loads the files, which should reduce memory consumption. I was able to load 3x large files than before.

Try the following patch and see if it's any better (this is against the most recent master):

```diff
diff --git a/src/editor/commands.lua b/src/editor/commands.lua
index 0ba9194..d52513a 100644
--- a/src/editor/commands.lua
+++ b/src/editor/commands.lua
@@ -61,16 +61,8 @@ function LoadFile(filePath, editor, file_must_exist, skipselection)
     end
   end
 
-  -- if not opened yet, try open now
-  local file_text = FileRead(filePath)
-  if file_text then
-    if GetConfigIOFilter("input") then
-      file_text = GetConfigIOFilter("input")(filePath,file_text)
-    end
-  elseif file_must_exist then
-    return nil
-  end
-
+  local filesize = FileSize(filePath)
+  if not filesize and file_must_exist then return nil end
   local current = editor and editor:GetCurrentPos()
   editor = editor or findUnusedEditor() or CreateEditor()
 
@@ -78,21 +70,29 @@ function LoadFile(filePath, editor, file_must_exist, skipselection)
   editor:SetupKeywords(GetFileExt(filePath))
   editor:MarkerDeleteAll(-1)
 
-  -- remove BOM from UTF-8 encoded files; store BOM to add back when saving
+  -- if not opened yet, try open now
+  local inputfilter = GetConfigIOFilter("input")
+  local file_text
+  editor:SetText("")
   editor.bom = string.char(0xEF,0xBB,0xBF)
-  if file_text and editor:GetCodePage() == wxstc.wxSTC_CP_UTF8
-  and file_text:find("^"..editor.bom) then
-    file_text = file_text:gsub("^"..editor.bom, "")
-  else
-    -- set to 'false' as checks for nil on wxlua objects may fail at run-time
-    editor.bom = false
-  end
-  editor:SetText(file_text or "")
+  FileRead(filePath, 1024*1024, function(s)
+      if not file_text then
+        -- remove BOM from UTF-8 encoded files; store BOM to add back when saving
+        if s and editor:GetCodePage() == wxstc.wxSTC_CP_UTF8 and s:find("^"..editor.bom) then
+          s = s:gsub("^"..editor.bom, "")
+        else
+          -- set to 'false' as checks for nil on wxlua objects may fail at run-time
+          editor.bom = false
+        end
+        file_text = s
+      end
+      editor:AppendText(inputfilter and inputfilter(filePath, s) or s)
+    end)
 
   -- check the editor as it can be empty if the file has malformed UTF8;
   -- skip binary files with unknown extensions as they may have any sequences;
   -- can't show them anyway.
-  if file_text and #file_text > 0 and #(editor:GetText()) == 0
+  if file_text and #file_text > 0 and editor:GetLength() == 0
   and (editor.spec ~= ide.specs.none or not IsBinary(file_text)) then
     local replacement, invalid = "\022"
     file_text, invalid = FixUTF8(file_text, replacement)
diff --git a/src/util.lua b/src/util.lua
index b0b0009..1a13371 100644
--- a/src/util.lua
+++ b/src/util.lua
@@ -299,7 +299,7 @@ end
 
 function FileSize(fname) return wx.wxFileExists(fname) and wx.wxFileSize(fname) or nil end
 
-function FileRead(fname, length)
+function FileRead(fname, length, callback)
   -- on OSX "Open" dialog allows to open applications, which are folders
   if wx.wxDirExists(fname) then return nil, "Can't read directory as file." end
 
@@ -307,6 +307,18 @@ function FileRead(fname, length)
   local file = wx.wxFile(fname, wx.wxFile.read)
   if not file:IsOpened() then return nil, wx.wxSysErrorMsg() end
 
+  if callback then
+    length = length or 4096
+    local pos = 0
+    while true do
+      local len, content = file:Read(length)
+      callback(content)
+      if len < length then break end
+      pos = pos + len
+      file:Seek(pos)
+    end
+    return true, wx.wxSysErrorMsg()
+  end
   local _, content = file:Read(length or wx.wxFileSize(fname))
   file:Close()
   return content, wx.wxSysErrorMsg()
```

**Comment by CapsAdmin** [comment]
It seems to get further. It shows the tab with the entire log but freezes right after and then crashes.

**Comment by pkulchenko** [comment]
I get about the same result on your file, which means that the fix is not good enough, but there is not much else I can do.

What I'm thinking about as an alternative, it to load a fixed chunk (let's say 1M) and only load more if you scroll to the end (I'm exploring how to detect that). The size of the default chunk can be configurable.

Would this help with your use case? You would not be able to load the entire file, but you'd still be able to see what's there at the beginning.

**Comment by CapsAdmin** [comment]
Sounds good. In my case I need to view the end of the log so there would have to be a way for me to goto to the end of the file. Maybe a special goto like ctrl-g and then type "inf" instead of the line number.

**Comment by pkulchenko** [comment]
Unfortunately, this may not help much as the file would still need to be read/loaded to the end. Maybe load the beginning and end and skip the middle part? This will look ugly though and will still require some way to load the middle part...

I don't see a good solution right now, but will ask in Scintilla forums. Another related issue is that the line numbering will be off if the file is partially loaded.

**Comment by pkulchenko** [comment]
Elias, I've incorporated several improvements we discussed (and some others), plus the changes based on my reading of Scintilla documentation (which includes pre-allocating of the space for loaded files). These should improve the handling of large files, although it may still fail on 300MB+ files. I also added showing a progress during loading. I'll continue checking, but I think this is the best I can do at this time.
